docs/design/jet/022-mongodb-connector.md

```diff
@@ -0,0 +1,345 @@
+        MongoDBSources.batch(
+                "batch-source",
+                "mongodb://127.0.0.1:27017",
+                "myDatabase",
+                "myCollection",
+                new Document("age", new Document("$gt", 10)),
+                new Document("age", 1)
+        );
+Pipeline p = Pipeline.create();
+BatchStage<Document> srcStage = p.readFrom(batchSource);
+```
+
+
+This example queries documents in a collection having the field
+`age` with a value greater than `10` and applies a projection so that only
+the `age` field is returned in the emitted document. User should be also
+able to add e.g. `MyUser.class` as the last argument, which will mean
+that all returned objects will be automatically mapped to Java POJO
+[see MongoDB docs](https://www.mongodb.com/developer/languages/java/java-mapping-pojos/)
+
+Example usage of sink:
+```
+Pipeline p = Pipeline.create();
+p.readFrom(Sources.list(list))
+ .map(i -> new Document("key", i))
+ .writeTo(
+     MongoDBSinks.mongodb(
+        "sink", 
+        "mongodb://localhost:27017",
+        "myDatabase",
+        "myCollection"
+     )
+ );
+```
+
+#### SQL Connector
+User should be able to run following SQL query:
+```sql
+CREATE MAPPING people
+TYPE MONGODB 
+OPTIONS (
+  'externalDataStoreRef'='mongodb-database' 
+)
+```
+
+In such cases, automatic schema inference will be used. User may also want
+to provide schema explicitly:
+
+```sql
+CREATE MAPPING people (
+    firstName VARCHAR(100),
+    lastName VARCHAR(100),
+    age INT
+)
+TYPE MONGODB 
+OPTIONS (
+  'externalDataStoreRef'='mongodb-database' 
+)
+```
+
+#### GenericMapStore support
+
+User will be able to configure map store:
+```
+Config config = new Config();
+config.addExternalDataStoreConfig(
+        new ExternalDataStoreConfig("mongodb-ref")
+            .setClassName(MongoDbDataStoreFactory.class.getName())
+            .setProperty("url", dbConnectionUrl)
+  );
+
+MapConfig mapConfig = new MapConfig(mapName);
+MapStoreConfig mapStoreConfig = new MapStoreConfig();
+mapStoreConfig.setClassName(GenericMapStore.class.getName());
+mapStoreConfig.setProperty(OPTION_EXTERNAL_DATASTORE_REF, "mongodb-ref");
+mapConfig.setMapStoreConfig(mapStoreConfig);
+instance().getConfig().addMapConfig(mapConfig);
+```
+
+### Technical Design
+
+#### Source and sink
+The `hazelcast-jet-contrib` repository contained MongoDB connector. It was
+based on the `SourceBuilder`, which is nice in general, but for more 
+flexibility we need to rework this connector to low-level Processor/
+ProcessorSupplier/ProcessorMetaSupplier:
+- `SqlConnector` class' methods are working on DAG level, so it will be easier
+  to reuse processors
+- custom MetaSupplier will make it easier to distribute workload based on
+  MongoDB's replica sets.
+
+We can check if we could use MongoDB's Replica sets. If yes, then
+our MetaSupplier can spawn as many ProcessorSuppliers as MongoDB has 
+replicas and configure Mongo client to use this specific replica. 
```

Comment:
Is it going to be also dependent on the number of the HZ member in the cluster?

Reply:
Yes - on each machine a Hazelcast member will be, we can check for replica sets available